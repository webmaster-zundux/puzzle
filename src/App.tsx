import "@fontsource-variable/open-sans";
import type { FC } from "react";
import { memo } from "react";
import "./styles.css";

import { PuzzleDataProvider } from "./components/PuzzleDataProvider";
import { PuzzlePage } from "./pages/PuzzlePage";

export const App: FC = memo(function App() {
  return (
    <PuzzleDataProvider>
      <PuzzlePage />
    </PuzzleDataProvider>
  );
});
