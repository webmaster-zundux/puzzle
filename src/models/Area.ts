import { Point } from "./Point";

export class Area {
  constructor(
    public p0: Point,
    public p1: Point,
    public p2: Point,
    public p3: Point,
  ) {}

  public static create({
    p0 = Point.getZeroPoint(),
    p1 = Point.getZeroPoint(),
    p2 = Point.getZeroPoint(),
    p3 = Point.getZeroPoint(),
  }: {
    p0: Point;
    p1: Point;
    p2: Point;
    p3: Point;
  }) {
    return new Area(p0, p1, p2, p3);
  }

  public static isPointWithinArea(point: Point | { x: number; y: number }, area: Area): boolean {
    const { p0, p2 } = area;
    const { x: px, y: py } = point;

    if (px < p0.x || px > p2.x) {
      return false;
    }

    if (py < p0.y || py > p2.y) {
      return false;
    }

    return true;
  }

  public static isPointWithinAreaByCornerPoints(
    point: Point | { x: number; y: number },
    areaPointP0: Point | { x: number; y: number },
    areaPointP2: Point | { x: number; y: number },
  ): boolean {
    const tx = point.x;
    const ty = point.y;

    const p0 = areaPointP0;
    const p2 = areaPointP2;

    if (tx < p0.x || tx > p2.x) {
      return false;
    }

    if (ty < p0.y || ty > p2.y) {
      return false;
    }

    return true;
  }

  public static doTwoAreasIntersect(areaA: Area, areaB: Area): boolean {
    const aP0 = areaA.p0;
    const aP2 = areaA.p2;

    const bP0 = areaB.p0;
    const bP2 = areaB.p2;

    return Area.doTwoAreasIntersectByCornerPoints(aP0, aP2, bP0, bP2);
  }

  public static doTwoAreasIntersectByCornerPoints(
    areaAPointP0: Point | { x: number; y: number },
    areaAPointP2: Point | { x: number; y: number },
    areaBPointP0: Point | { x: number; y: number },
    areaBPointP2: Point | { x: number; y: number },
  ): boolean {
    const aP0 = areaAPointP0;
    const aP2 = areaAPointP2;

    const bP0 = areaBPointP0;
    const bP2 = areaBPointP2;

    let intersectionOnXAxis = true;
    let intersectionOnYAxis = true;

    if (aP2.x < bP0.x || aP0.x > bP2.x) {
      intersectionOnXAxis = false;
    }

    if (aP2.y < bP0.y || aP0.y > bP2.y) {
      intersectionOnYAxis = false;
    }

    return intersectionOnXAxis && intersectionOnYAxis;
  }

  public static getDefaultArea() {
    return new Area(new Point(0, 0), new Point(1, 0), new Point(1, 1), new Point(0, 1));
  }

  toString() {
    return `Area {p0: ${this.p0}, p1: ${this.p1}, p2: ${this.p2}, p3: ${this.p3}}`;
  }
}
