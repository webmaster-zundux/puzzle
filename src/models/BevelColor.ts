export const BEVEL_EFFECT_HIGHLIGHT = "highlight" as const;
export const BEVEL_EFFECT_SHADOW = "shadow" as const;

export type BEVEL_EFFECT_TYPE = typeof BEVEL_EFFECT_SHADOW | typeof BEVEL_EFFECT_HIGHLIGHT;
