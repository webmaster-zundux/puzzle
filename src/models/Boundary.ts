export class Boundary {
  constructor(
    public x: number,
    public y: number,
    public width: number,
    public height: number,
  ) {}

  public static getIntersectionBoundaryOfTwoBoundaries(a: Boundary, b: Boundary): Boundary | undefined {
    const x1 = a.x > b.x ? a.x : b.x;
    const y1 = a.y > b.y ? a.y : b.y;

    const aP2X = a.x + a.width;
    const aP2Y = a.y + a.height;

    const bP2X = b.x + b.width;
    const bP2Y = b.y + b.height;

    const x2 = aP2X < bP2X ? aP2X : bP2X;
    const y2 = aP2Y < bP2Y ? aP2Y : bP2Y;

    const width = x2 - x1;
    const height = y2 - y1;

    // checking if they are intersect
    if (width <= 0 || height <= 0) {
      return undefined;
    }

    return {
      x: x1,
      y: y1,
      width,
      height,
    };
  }
}
