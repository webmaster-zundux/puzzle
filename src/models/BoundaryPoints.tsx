import type { Point } from "./Point";

export type BoundaryPoints = {
  tl: Point;
  br: Point;
};
