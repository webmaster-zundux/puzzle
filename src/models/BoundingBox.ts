export class BoundingBox {
  constructor(
    public x: number = 0,
    public y: number = 0,
    public width: number = 1,
    public height: number = 0,
  ) {}

  public static create({ x, y, width, height }: { x: number; y: number; width: number; height: number }) {
    return new BoundingBox(x, y, width, height);
  }
}
