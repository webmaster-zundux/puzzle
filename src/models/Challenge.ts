import type { Id } from "../core/puzzle/BaseEntity";
import { uuid } from "../utils/uuid";

export type ChallengeId = Id;

const generateChallengeIdFunction = () => {
  const newUUID = uuid();
  return newUUID;
};

const EXPOSE_CHALLENGE_ID_GENERATION_TO_WINDOW_OBJECT =
  import.meta.env.VITE_TESTING_CHALLENGE_ID_GENERATION_EXPOSE_TO_WINDOW_OBJECT === "1";
if (EXPOSE_CHALLENGE_ID_GENERATION_TO_WINDOW_OBJECT) {
  // @ts-expect-error property 'generateChallengeId' does not exist on type 'Window & typeof globalThis'
  if (typeof window?.generateChallengeId !== "function") {
    Object.defineProperty(window, "generateChallengeId", {
      value: generateChallengeIdFunction,
      writable: false,
      configurable: false,
    });
  }
}

// @ts-expect-error property 'generateChallengeId' does not exist on type 'Window & typeof globalThis'
export const generateChallengeId = window?.generateChallengeId ?? generateChallengeIdFunction;
