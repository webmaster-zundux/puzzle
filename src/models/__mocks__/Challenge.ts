export const generateChallengeId = () => {
  const newUUID = "364bf6d0-6686-4892-be7d-618ac2fcdcc7";
  return newUUID;
};
