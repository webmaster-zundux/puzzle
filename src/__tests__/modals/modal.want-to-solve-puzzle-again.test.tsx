import { screen, waitFor, within } from "@testing-library/react";
import userEvent from "@testing-library/user-event";
import { vi } from "vitest";
import { App } from "../../App";
import * as Challenge from "../../models/Challenge";
import { renderIntoDocumentBody } from "../__tests-utils__/renderIntoDocumentBody";
import {
  isSolvePuzzleAgainDialogOpened,
  openSolvePuzzleAgainDialogByClickOnSolvePuzzleAgainButtonInSidebar,
} from "./modal-helpers";

describe(`modal dialog "want to solve the puzzle again"`, () => {
  describe(`changes it's visibility state`, () => {
    it(`opens when clicking on "restart puzzle" button in the sidebar`, async () => {
      const user = userEvent;
      renderIntoDocumentBody(<App />);
      const sidebar = screen.getByRole("complementary", { name: /sidebar/i });
      expect(screen.queryByRole("dialog")).not.toBeInTheDocument();
      await user.click(screen.getByRole("application", { name: /puzzle view/i }));
      await waitFor(() => {
        expect(within(sidebar).getByRole("button", { name: /restart puzzle/i })).toBeEnabled();
      });

      await user.click(within(sidebar).getByRole("button", { name: /restart puzzle/i }));

      await isSolvePuzzleAgainDialogOpened({ isShortCheck: false });
    });

    it(`opens when clicking on "solve puzzle again" button in the sidebar`, async () => {
      renderIntoDocumentBody(<App />);

      await openSolvePuzzleAgainDialogByClickOnSolvePuzzleAgainButtonInSidebar({ isShortCheck: false });
    });

    it("closes when clicking on the dialog backdrop", async () => {
      const user = userEvent;
      const { asFragment } = renderIntoDocumentBody(<App />);
      const renderWithoutDialog = asFragment();
      const dialog = await openSolvePuzzleAgainDialogByClickOnSolvePuzzleAgainButtonInSidebar();
      const backdrop = within(dialog).getByLabelText(/modal-backdrop/i);
      expect(backdrop).toBeInTheDocument();

      await user.click(backdrop);

      expect(screen.queryByRole("dialog")).not.toBeInTheDocument();
      expect(renderWithoutDialog).toMatchDiffSnapshot(asFragment(), undefined, `without dialog`);
    });

    it('closes when pressing on "esc" key', async () => {
      const user = userEvent;
      renderIntoDocumentBody(<App />);
      await openSolvePuzzleAgainDialogByClickOnSolvePuzzleAgainButtonInSidebar();

      await user.keyboard("{Escape}");

      expect(screen.queryByRole("dialog")).not.toBeInTheDocument();
    });

    it('closes when clicking on "cancel" button', async () => {
      const user = userEvent;
      vi.spyOn(Challenge, "generateChallengeId")
        .mockImplementationOnce(() => "b47663e4-95ab-4472-9344-0f165a6a7ff9")
        .mockImplementationOnce(() => "89e80c0c-186a-46db-b242-b147c88a2a3e");
      renderIntoDocumentBody(<App />);
      expect(Challenge.generateChallengeId).toHaveBeenCalledOnce();
      const initialChallengeSeed = screen.getByRole("note", { name: /challenge seed:/i }).textContent;
      expect(initialChallengeSeed).toBe("b47663e4-95ab-4472-9344-0f165a6a7ff9");
      expect(screen.queryByRole("dialog")).not.toBeInTheDocument();
      const dialog = await openSolvePuzzleAgainDialogByClickOnSolvePuzzleAgainButtonInSidebar({ user });

      await user.click(within(dialog).getByRole("button", { name: /cancel/i }));

      expect(Challenge.generateChallengeId).toHaveBeenCalledOnce();
      expect(screen.queryByRole("dialog", { name: /want to solve the puzzle again\?/i })).not.toBeInTheDocument();
      const challengeSeedAfterCanceledRestartOfPuzzle = screen.getByRole("note", {
        name: /challenge seed:/i,
      }).textContent;
      expect(challengeSeedAfterCanceledRestartOfPuzzle).toBe(initialChallengeSeed);
    });

    it('closes and creates a new puzzle challenge (changes only pieces positions and the cutting of the pieces) when clicking on "solve again" button in the dialog', async () => {
      const user = userEvent;
      vi.spyOn(Challenge, "generateChallengeId")
        .mockImplementationOnce(() => "b47663e4-95ab-4472-9344-0f165a6a7ff9")
        .mockImplementationOnce(() => "89e80c0c-186a-46db-b242-b147c88a2a3e");
      const { asFragment } = renderIntoDocumentBody(<App />);
      const firstRenderWithoutDialog = asFragment();
      const initialChallengeSeed = screen.getByRole("note", { name: /challenge seed:/i }).textContent;
      expect(Challenge.generateChallengeId).toHaveBeenCalledOnce();
      expect(initialChallengeSeed).toBe("b47663e4-95ab-4472-9344-0f165a6a7ff9");
      const dialog = await openSolvePuzzleAgainDialogByClickOnSolvePuzzleAgainButtonInSidebar({ user });

      await user.click(within(dialog).getByRole("button", { name: /solve again/i }));

      expect(Challenge.generateChallengeId).toHaveBeenCalledTimes(2);
      expect(screen.queryByRole("dialog", { name: /want to solve the puzzle again\?/i })).not.toBeInTheDocument();
      const secondChallengeSeed = screen.getByRole("note", { name: /challenge seed:/i }).textContent;
      expect(secondChallengeSeed).not.toBe(initialChallengeSeed);
      expect(firstRenderWithoutDialog).toMatchDiffSnapshot(asFragment(), undefined, "the second puzzle challenge");
    });
  });
});
