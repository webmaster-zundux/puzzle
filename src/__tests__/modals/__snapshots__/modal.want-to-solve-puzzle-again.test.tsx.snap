// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`src/__tests__/modals/modal.want-to-solve-puzzle-again.test.tsx > modal dialog "want to solve the puzzle again" > changes it's visibility state > closes and creates a new puzzle challenge (changes only pieces positions and the cutting of the pieces) when clicking on "solve again" button in the dialog: the second puzzle challenge 1`] = `
"Snapshot Diff:
- First value
+ Second value

@@ -447,11 +447,11 @@
            </div>
            <div
              aria-labelledby="challenge-seed-label"
              role="note"
            >
-             b47663e4-95ab-4472-9344-0f165a6a7ff9
+             89e80c0c-186a-46db-b242-b147c88a2a3e
            </div>
          </div>
        </aside>
      </main>
    </div>"
`;

exports[`src/__tests__/modals/modal.want-to-solve-puzzle-again.test.tsx > modal dialog "want to solve the puzzle again" > changes it's visibility state > closes when clicking on the dialog backdrop: without dialog 1`] = `
"Snapshot Diff:
Compared values have no visual difference."
`;
